modules/services/xsettingsd.nix

```diff
@@ -21,29 +33,68 @@ in {
           Package containing the <command>xsettingsd</command> program.
         '';
       };
-    };
-  };
 
-  config = mkIf cfg.enable {
-    assertions = [
-      (lib.hm.assertions.assertPlatform "services.xsettingsd" pkgs
-        lib.platforms.linux)
-    ];
-
-    systemd.user.services.xsettingsd = {
-      Unit = {
-        Description = "xsettingsd";
-        After = [ "graphical-session-pre.target" ];
-        PartOf = [ "graphical-session.target" ];
+      settings = mkOption {
+        type = with types; attrsOf (oneOf [ bool int str ]);
+        default = { };
+        example = literalExample ''
+          {
+            "Net/ThemeName" = "Numix";
+            "Xft/Antialias" = true;
+            "Xft/Hinting" = true;
+            "Xft/RGBA" = "rgb";
+          }
+        '';
+        description = ''
+          Xsettingsd options for configuration file. See
+          <link xlink:href="https://github.com/derat/xsettingsd/wiki/Settings"/>
+          for documentation on these values.
+        '';
       };
 
-      Install.WantedBy = [ "graphical-session.target" ];
-
-      Service = {
-        Environment = "PATH=${config.home.profileDirectory}/bin";
-        ExecStart = "${cfg.package}/bin/xsettingsd";
-        Restart = "on-abort";
+      configLocation = mkOption {
```

Comment:
I think you can omit this option and instead do something like

```
ExecStart =
  let
    configFile = pkgs.writeText "xsettingsd.conf" (renderSettings cfg.settings + "\n");
  in toString (["${cfg.package}/bin/xsettingsd] ++ optional (cfg.settings != {}) "-c ${configFile}";
```

This would also make the `mkMerge` and `onChange` unnecessary.

Reply:
This makes sense, but i think then (without `onChange`) the service would have to be manually reloaded?

And also, i think actually writting to the expected location could be helpful (for inspecting, or in case some other program/script greps it)?

Reply:
And i think this wouldn't work for someone who is using just `enable` and `configLocation`, but not `settings`

Reply:
Alternatively, to allow easier testing of the generated file you could add an option along the line of

```
configFile = mkOption {
  type = types.nullOr types.package;
  internal = true;
  readOnly = true;
  default =
    if cfg.settings == {} then
      null
    else
      pkgs.writeText "xsettingsd.conf" (renderSettings cfg.settings + "\n");
};
```

then have

```
ExecStart = toString (["${cfg.package}/bin/xsettingsd]
     ++ optional (cfg.configFile != null) "-c ${cfg.configFile}");
```

the test should then be able to use

```
assertFileContent ${config.services.xsettings.configFile} ${./basic-configuration.conf}
```

At least I think it should work. Everything above is completely untested 😉

Reply:
Did some quick testing here. We would still need some way to `SIGHUP` (or restart it altogether) xsettingsd, as it doesn't watch the configuration file.

Also, rendering the text by using `default` would break the usecase of setting both `settings` and `configLocation`, in case someone wishes to use home-manager to manage it _and_ have it somewhere else other than ~.

Reply:
No need to reload the service, the systemd module will make sure it is restarted.

If other programs are using the config location then we can't change it here anyway since they most likely would expect the default location. But I don't think any other program is using this file directly and if you want to look up a setting it would be best to use the `dump_xsettings` tool that is provided by `xsettingsd`. So I think it should be safe to simply omit the file from the home directory and instead refer to it directly in the Nix store.

If somebody really want to see the content of it then they can always get the path using

    systemctl --user cat xsettingsd.service

Reply:
Well, to be fair, automatic service restart will only happen if the user have [`systemd.user.startServices`](https://nix-community.github.io/home-manager/options.html#opt-systemd.user.startServices) set to `legacy` or `sd-switch`. But even will `suggest` the user should be prompted to restart the service due to its content changing.

Reply:
Ooh that's why they weren't restarting for me. That makes sense.

I've since implemented the changes but they seem to fail testing with a cryptic-ish message, i'm gonna go ahead and push them so you can take a look at it while i sort it out.

Reply:
I'm... Not really sure what could be causing the test to fail
```
~~> nix-shell --pure tests -A run.xsettingsd-basic-configuration
error (ignored): error: bad archive: input doesn't look like a Nix archive
error: getting status of '/home/misterio/Documents/Programacao/Pessoal/home-manager/modules/services/cfg.package/bin/xsettingsd': No such file or directory
(use '--show-trace' to show detailed location information)
```

Reply:
Oh of course, i closed the curly brackets wrongly, fixed now.